src/compiler/tsbuild.ts

```diff
@@ -0,0 +1,1222 @@
+            reason: string;
+        }
+
+        /**
+         * This project doesn't have any outputs, so "is it up to date" is a meaningless question.
+         */
+        export interface ContainerOnly {
+            type: UpToDateStatusType.ContainerOnly;
+        }
+
+        /**
+         * The project is up to date with respect to its inputs.
+         * We track what the newest input file is.
+         */
+        export interface UpToDate {
+            type: UpToDateStatusType.UpToDate | UpToDateStatusType.UpToDateWithUpstreamTypes;
+            newestInputFileTime: Date;
+            newestInputFileName: string;
+            newestDeclarationFileContentChangedTime: Date;
+            newestOutputFileTime: Date;
+            newestOutputFileName: string;
+            oldestOutputFileName: string;
+        }
+
+        /**
+         * One or more of the outputs of the project does not exist.
+         */
+        export interface OutputMissing {
+            type: UpToDateStatusType.OutputMissing;
+            /**
+             * The name of the first output file that didn't exist
+             */
+            missingOutputFileName: string;
+        }
+
+        /**
+         * One or more of the project's outputs is older than its newest input.
+         */
+        export interface OutOfDateWithSelf {
+            type: UpToDateStatusType.OutOfDateWithSelf;
+            outOfDateOutputFileName: string;
+            newerInputFileName: string;
+        }
+
+        /**
+         * This project depends on an out-of-date project, so shouldn't be built yet
+         */
+        export interface UpstreamOutOfDate {
+            type: UpToDateStatusType.UpstreamOutOfDate;
+            upstreamProjectName: string;
+        }
+
+        /**
+         * This project depends an upstream project with build errors
+         */
+        export interface UpstreamBlocked {
+            type: UpToDateStatusType.UpstreamBlocked;
+            upstreamProjectName: string;
+        }
+
+        /**
+         * One or more of the project's outputs is older than the newest output of
+         * an upstream project.
+         */
+        export interface OutOfDateWithUpstream {
+            type: UpToDateStatusType.OutOfDateWithUpstream;
+            outOfDateOutputFileName: string;
+            newerProjectName: string;
+        }
+    }
+
+    interface FileMap<T> {
+        setValue(fileName: string, value: T): void;
+        getValue(fileName: string): T | never;
+        getValueOrUndefined(fileName: string): T | undefined;
+        hasKey(fileName: string): boolean;
+        removeKey(fileName: string): void;
+        getKeys(): string[];
+    }
+
+    /**
+     * A FileMap maintains a normalized-key to value relationship
+     */
+    function createFileMap<T>(): FileMap<T> {
+        // tslint:disable-next-line:no-null-keyword
+        const lookup: { [key: string]: T } = Object.create(/*prototype*/ null);

```

Comment:
createMap?

Reply:
ES Maps has better memory characteristics than objects, since they do not hold on to the property name strings.